controller/volume_controller.go

```diff
@@ -904,17 +910,27 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 			if !dataExists {
 				log.Warn("Cannot auto salvage volume: no data exists")
 			} else {
+				// This salvage is for revision counter enabled case
 				salvaged := false
+				// Since all replica failed, mark engine controller salvage requested
+				e.Spec.SalvageRequested = true
+				logrus.Infof("All replicas are failed, set engine salvageRequested to %v", e.Spec.SalvageRequested)
+
+				// Bring up the replicas for auto-salvage
 				for _, r := range failedUsableReplicas {
 					if util.TimestampWithinLimit(lastFailedAt, r.Spec.FailedAt, AutoSalvageTimeLimit) {
 						r.Spec.FailedAt = ""
-						log.WithField("replica", r.Name).Warn("Automatically salvaged volume replica")
-						msg := fmt.Sprintf("Replica %v of volume %v has been automatically salvaged", r.Name, v.Name)
-						vc.eventRecorder.Event(v, v1.EventTypeWarning, EventReasonAutoSalvaged, msg)
-						salvaged = true
+						log.WithField("replica", r.Name).Warn("Automatically salvaging volume replica")
+						// Revision counter enabled case, engine will
+						// check revision counter of replicas
+						if v.Spec.RevisionCounterDisabled {
```

Comment:
Why do we need this check? Then `salvaged` cannot be `true` if the revision counter is enabled. That doesn't make sense.

Reply:
Sorry, should be `if !v.Spec.RevisionCounterDisabled {`. Will update. Thanks.